### ISSUE TITLE
n-dimensional Properties

### ISSUE THREAD

**Issue by brisvag** [napari/napari#2917]
This issue is part of #2866, exploring the possibility of n-dimensional properties.

# Goals
The main goal of this issue is to find a way (if it's possible) to extend the n-dimensionality of napari to layer properties. I will focus mainly on examples with `Points`, since they are probably the easiest application, but this should be applicable to all the other layers with properties.

A thing to keep in mind that we might want from this is what was dicussed in #2592: we not only would like the ability to broadcast properties to the data's shape, but also the other way around.

# Options
So far I've been exploring the idea of using [xarray Datasets](http://xarray.pydata.org/en/stable/user-guide/data-structures.html#dataset) as an n-dimensional version of pandas Dataframes. Both pandas and xarray devs [seem to agree on this](https://github.com/pandas-dev/pandas/issues/12940), and there's been some [collaboration between the two libraries](http://xarray.pydata.org/en/stable/user-guide/pandas.html#transitioning-from-pandas-panel-to-xarray). A lot of the expected functionality is there (indexing by column names, broadcasting and so on), though there *are* differences here and there.

An alternative is to use [pandas MultiIndexing](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html), as suggested in a few places [like here](https://stackoverflow.com/questions/36760414/how-to-create-pandas-dataframes-with-more-than-2-dimensions).

I currently lean more towards xarray, since it feels more intuitive and more geared towards this. It also has some super cool [interpolation logic](http://xarray.pydata.org/en/stable/user-guide/interpolation.html#interpolating-data) that could go a long way in napari (you could have properties for beginning and end of a time series and easily interpolate them at each time step!). I also should [interface with dask easily](http://xarray.pydata.org/en/stable/user-guide/dask.html?highlight=dask#using-dask-with-xarray).

# Implementation
[I won't be using napari for now, @andy-sweet is taking care of that side of things :)]

Let's say we want 5 points in 3D space with 2 properties: `color` and `size` (that we could then link to the homonymous representation). We want position and color to be static, but size change over time.
```python
import numpy as np

# five 3D points + time with 20 time steps are defined as 100 different coordinates
points = np.random.rand(5, 4)
points = np.tile(points, (20, 1))
points[:, 0] = np.arange(20).repeat(5)

# we explicitly set the properties for all the points at all time steps
color = np.random.choice(['blue', 'red'], 5)
color = np.tile(color, 20)
size = np.random.rand(100)

properties = {'color': color, 'size'=size}
...
```
As evidenced by the use of `np.tile`, we had to duplicate both the data along some dimension (in this case time) and the `color` property.

Here's an idea on how one could do this in xarray.
```python
import numpy as np
import xarray as xr

# five 3D points, static
points = np.random.rand(5, 3)

# only size needs the extra dim
color = np.random.choice(['blue', 'red'], 5)
# this time size is defined for the 5 points over 20 time steps
size = np.random.rand(5, 20)

# create a dataset and assing the properties using `idx` as index coordinate
ds = xr.Dataset()
ds = ds.assign({'color': (['idx'], color)})
# size has an additional coordinate, let's call it 'time' for now (more on it later)
ds = ds.assign({'size': (['idx', 'time'], size)})

# we can now slice these properties based on index to get only some points
ds.isel(idx=slice(1, 4, 1))
# or based on time, to get it at each timestep
ds.isel(time=2)
# or as I mentioned earlier, even interpolate! (only size in this case)
ds.interp(time=2.3)

# we can also simply access each property and index into it the good old way:
ds.size.shape  # (5, 20)
ds.size[:, -5:] # last 5 time steps
#  you can always access the underlying np.ndarray with
ds.size.values
```

# Problems
Something that you might have noticed is that I'm not sure how this should be handled in terms of interfacing named dimensions with napari's pure integer indexing. It might be the trickiest bit. I'm posting this issue early in hopes of feedback on this particular problem! Points to consider:
- xarray does not accept integers as dimension names
- dimension order is not guaranteed (*I think*), so working directly with the underlying numpy arrays might be risky

Things get more complicated when considering how dimensions in something like `Points` are technically *coordinates* rather than *dimensions*, which clashes with this xarray approach. For example if we want to change the above example to have point coordinates change over time, we have to express it as 100 different 4D points, where the 0th dimension encodes time (just like the first example):

```python
points = np.random.rand(5, 4)
points = np.tile(points, (20, 1))
points[:, 0] = np.arange(20).repeat(5)
```

Now, how do we broadcast this to the `time` dimension in the previous example? We need to "unstack" (unsure on the terminology here) this dimension into an *actual* dimension, so that the shape of our points is actually `(5, 3, 20)` rather than `(100, 4)`.
This is not very nice, cause we lose napari's agnosticism about dimensions. Ideally, then, we would want to completely unstack into a `(5, x, y, z, 20)`, but then we lose the sparseness of points and end up with a dense image instead. I'm sure I'm missing something ovious here.

---

I would love to hear some thoughts and opinions on the use of xarray, and some suggestions on the dimension naming issues!

**Comment by brisvag** [comment]
One problem that I hadn't considered so far is: what happens if the number of points is different in each slice of the 4th dimension? Broadcasting simply can't work there.

**Comment by andy-sweet** [comment]
Thanks for this exploration and sorry for the slow feedback.

Not sure if this will help your problems or add to them, but here are some (rambling) thoughts on dimensions/coordinates.

It seems like you want to coerce sparsely defined points into a dense rectilinear image-like grid. That seems meaningful for things like rendering, but not so much for data representation because it's a lossy operation. Instead, why not think about the other direction? If you have an image that is defined as a dense rectilinear grid, consider it as a set of sparsely defined pixels instead.

For example, let's take a 2D RGBA image. In napari, this will likely be backed by a 3-dimensional numpy array of shape (Y, X, 4). But you could also represent it as two 2-dimensional arrays of shapes (N, 2) and (N, 4), where N is a pixel index dimension. The (N, 2) array encodes the (x, y) location of each pixel whereas the (N, 4) array encodes the (r, g, b, a) color of each pixel. This definition is consistent with that of points (e.g. a 2D points layer has an (N, 2) array to encode location and several Nx? arrays to encode different properties like size, color), so we should be able to use the same underlying representation. It also makes images fit better into more general data visualization frameworks [like Altair](https://uwdata.github.io/visualization-curriculum/altair_marks_encoding.html#).

The obvious downside is that the general sparse representation of an image is memory inefficient when the image pixels really are a rectilinear grid of samples. It can also be inconvenient and time inefficient when performing certain operations (e.g. slicing through one dimension, adding two images, interpolation). For these reasons, I doubt we actually want to use this representation for images in napari.

The other important thing here is that some dimensions are different, which I think you already touched upon.

First, some dimensions have ordinal data types where the order defined is meaningful for that dimension (e.g. spatial x-dimension, time), whereas other have nominal data types (e.g. color defined as strings like 'red', 'cyan', where order can be imposed, but it's arbitrary). I think index is more of a nominal type because while you can order indices, that order is not always meaningful (though for dense images it kind of is).

Second, I tend to think of images as functions that map from a domain to a range. Typically, I think of spatio-temporal dimensions as defining the domain of an image and other dimensions as defining the range (e.g. the RGBA image maps from a 2D location to a color) because that allows me to do things like spatio-temporal interpolation or transformation very easily. Sometimes other dimensions may also fit into the domain (e.g. angle, direction), but often the important thing is whether the dimensions in the domain vary meaningfully together (i.e. are they meaningful co-ordinates?). For example, we could put color into the domain of the RGBA image, but the x-color slice is not as meaningful to look at (e.g. how do we interpolate between the R and G values in the slice?). I think this is where napari's dimensional agnosticism causes some tension - at least in my mind.

**Comment by jni** [comment]
> Now, how do we broadcast this to the time dimension in the previous example? We need to "unstack" (unsure on the terminology here) this dimension into an actual dimension, so that the shape of our points is actually (5, 3, 20) rather than (100, 4).

Yeah I don't think we want to unstack here. I think a key issue is that if you are thinking of points as individual data points that vary over a dimension (including in coordinate), then we are back to (say) 4D coordinates *together* with a point index in the properties. This would actually bring it nicely in parity with the Labels layer.

However, the bigger framework shift that I think we might need to make is that points are in some sense discrete/categorical and so broadcasting is perhaps better considered in some *database join* kind of framework rather than in NumPy's dimension broadcasting framework. But I'm not so well equipped to deal with that one...

> For example, we could put color into the domain of the RGBA image, but the x-color slice is not as meaningful to look at (e.g. how do we interpolate between the R and G values in the slice?). I think this is where napari's dimensional agnosticism causes some tension - at least in my mind.

Yeah, this is correct @andy-sweet and in this case the channel dimension is indeed special — channels should be layers. We only fall back on displaying them in a slider when we don't have that information.
